File: c#/Lab14/Lab14/Lab14_4/City.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace Lab14_4
{
    class City
    {
        /*Місто:
         * 1. Назва
         * 2. Кількість населення (метод, який виводить статус міста)
         * 3. Мер
         * 4. Дата заснування (метод, який виводить, скільки років місту на даний момент)
         * 5. Геолокація (країна, в якій знаходиться місто)
         * 6. Історія міста
         * 7. Визначні місця
         */
        public string Name { get; set; }
        public string Mayor { get; set; }
        public int YearOfFounding { get; set; }
        public string History { get; set; }
        public string Geography { get; set; }
        public uint Population { get; set; }
        public MyList<string> CultureArea { get; set; }

        public City()
        {

        }

        public City(string name, string mayor, int year, string history, string geography, uint population, MyList<string> culture)
        {
            this.Name = name;
            this.Mayor = mayor;
            this.YearOfFounding = year;
            this.History = history;
            this.Geography = geography;
            this.Population = population;
            this.CultureArea = culture;
        }

        public void InputInfo()
        {
            Console.Write("Input city name : ");
            this.Name = Console.ReadLine();
            Console.Write("\nInput mayor : ");
            this.Mayor = Console.ReadLine();
            while (true)
            {
                Console.Write("\nInput year of founding : ");
                var i = 0;
                try
                {
                    this.YearOfFounding = int.Parse(Console.ReadLine());
                    i++;
                }
                catch (FormatException e)
                {
                    Console.WriteLine("Value error! Change your data!");
                }
                if (this.YearOfFounding < 0)
                {
                    Console.WriteLine("Range error! Change your data!");
                    i = 0;
                }
                if (i != 0) break;
            }
            Console.Write("\nInput city history : ");
            this.History = Console.ReadLine();
            Console.Write("\nInput location of city : ");
            this.Geography = Console.ReadLine();
            while (true)
            {
                var i = 0;
                uint n = 0;
                Console.Write("\nWrite total population : ");
                if (!uint.TryParse(Console.ReadLine(), out n))
                {
                    Console.WriteLine("Value error!");
                }
                else
                {
                    this.Population = n;
                    break;
                }
            }
            Console.WriteLine("\nInput your culture (type \"exit\" to stop) :");
            var culture = new MyList<string>();
            while (true)
            {
                string temp = Console.ReadLine();
                if (temp == "exit") break;
                culture.Add(temp);
            }
            this.CultureArea = culture;
        }
        public void GetInfo()
        {
            Console.WriteLine($"City : {this.Name}\nMayor : {this.Mayor}\nYear of founding : {this.YearOfFounding}\nCity history : {this.History}" +
                $"\nCity location : {this.Geography}\nThe total population : {this.Population}\nSome culture of city :\n");
            var i = 1;
            foreach (var c in this.CultureArea)
            {
                Console.WriteLine($"{i}. {c}");
                i++;
            }
        }
        private int GetCityAge()
        {
            DateTime now = DateTime.Now;
            int age = now.Year - this.YearOfFounding;
            //if (age % 5 == 0)
            //{
            //    Console.WriteLine($"Jubilee! City age is {age}!");
            //}
            //else
            //{
            //    Console.WriteLine($"City age is {age}!");
            //}
            return age;
        }
        public void GetCityStatus()
        {
            string status = "";
            if (this.Population < 50000)
            {

                status = "small";
            }
            else if (this.Population < 100000)
            {

                status = "medium";
            }
            else if (this.Population < 500000)
            {

                status = "large";
            }
            else if (this.Population < 1000000)
            {

                status = "very large";
            }
            else
            {
                status = "millionaire";
            }
            Console.WriteLine($"City {this.Name} is {status}");
        }
        public static void GetOldestCity(MyList<City> cities)
        {
            int minAge = cities[0].YearOfFounding;
            var k = 0;
            for (var i = 0; i < cities.Count; i++)
            {
                if (cities[i].YearOfFounding < minAge)
                {
                    minAge = cities[i].YearOfFounding;
                    k = i;
                }
            }
            int cityAge = cities[k].GetCityAge();
            Console.WriteLine($"The oldest city is {cities[k].Name}. Its age is : {cityAge}");
        }
        //private string GetFoundingStatus()
        //{
        //    string status = "";
        //    int year = GetCityAge();
        //    if 
        //    return status;
        //}
    }
}

File: python/Lab10/city.py
import datetime
def correct_float(string):
    while True:
        try:
            point = float(input("{}".format(string)))
        except ValueError:
            print("Value error! Try again!\n")
        else:
            if point >= 0:
                return point
            print("Range error! Try again!\n")

class City(object):
    
    def __init__(self, name = None, mayor = None, year_of_founding = None, history = None, geography = None, population = None, culture = []) -> None:
        self.name = name
        self.mayor = mayor
        self.year_of_founding = year_of_founding
        self.history = history
        self.geography = geography
        self.population = population
        self.culture_area = culture
    
    @property
    def get_name(self):
        return self.name
    @property
    def get_mayor(self):
        return self.mayor
    @property
    def get_year_of_founding(self):
        return self.year_of_founding
    @property
    def get_history(self):
        return self.history
    @property
    def get_geography(self):
        return self.geography
    @property
    def get_population(self):
        return self.population
    @property
    def get_culture(self):
        return self.culture_area

    @property
    def set_name(self, value):
        self.name = value
    @property
    def set_mayor(self, value):
        self.mayor = value
    @property
    def set_year_of_founding(self, value):
        self.year_of_founding = value
    @property
    def set_history(self, value):
        self.history = value
    @property
    def set_geography(self, value):
        self.geography = value
    @property
    def set_population(self, value):
        self.population = value
    @property
    def set_culture(self, value):
        self.culture_area = value

    def input_info(self):
        self.name = input("Input city name : ")
        self.mayor = input("Input city mayor : ")
        while True:
            try:
                self.year_of_founding = int(input("Input year of founding : "))
            except ValueError:
                print("Value error! Change your data!")
            else:
                if (self.year_of_founding >= 0): break
                else:
                    print("Range error! Change your data!")
        self.history = input("Input city history : ")
        self.geography = input("Input city location : ")
        while True:
            try:
                self.population = int(input("Input total population : "))
            except ValueError:
                print("Value error! Change your data!")
            else:
                if (self.population >= 0): break
                else:
                    print("Range error! Change your data!")
        print("Input culture (type \"exit\" to break)")
        while True:
            data = input()
            if data == "exit": break
            self.culture_area.append(data)

    
    def get_info(self):
        print(f"City : {self.name}\nMayor : {self.mayor}\nYear of founding : {self.year_of_founding}\nCity history : {self.history}\nCity location : {self.geography}\nThe total population : {self.population}\nSome culture of city :\n")
        i = 1
        for item in self.culture_area:
            print(f"{i}. {item}")
        print("\n")

    def city_age(self):
        now = datetime.datetime.today().year
        age = now - self.year_of_founding
        return age
    
    def get_city_status(self):
        status = ""
        if self.population < 50000:
                status = "small"
        elif self.population < 100000:
            status = "medium"
        elif self.population < 500000:
            status = "large"
        elif self.population < 1000000:
            status = "very large"
        else:
            status = "millionaire"
        print(f"City {self.name} is {status} city")
    
    def get_oldest_city(cities):
        min_age = cities[0].year_of_founding
        k = 0
        for i in range(len(cities)):
            if cities[i].year_of_founding < min_age:
                min_age = cities[i].year_of_founding
                k = i
        city_age = cities[k].city_age()
        print(f"The oldest city is {cities[k].name}. Its age is : {city_age}")
